torch/functional.py

```diff
@@ -98,28 +98,38 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
         >>> # can recover A from factorization
         >>> A_ = torch.bmm(P, torch.bmm(A_L, A_U))
     """
-
-    sz = LU_data.size(-1)
-
+    shape = LU_data.shape
+    # In generalized LU factorization, the following shape relations hold:
+    #   A.shape[-2:] == (m, n)
+    #   P.shape[-2:] == (m, m)
+    #   L.shape[-2:] == (m, k)
+    #   U.shape[-2:] == (k, n)
+    # where k = min(m, n)
+    m, n = shape[-2:]
+    k = min(m, n)
     if unpack_data:
         U = LU_data.triu()
+        if m != k:
+            U = U[..., :k, :]
```

Comment:
I was going to suggest ```torch.narrow(U, -2, 0, k)```, but your version is more readable.

Reply:
Would `narrow` be preferred performance-wise?

Reply:
Seems like it:
```python
In [1]: import torch                                                                                                 

In [2]: U = torch.randn(1000, 100, 100)                                                                              

In [3]: %%timeit 
   ...: U[..., :50, :] 
   ...:  
   ...:                                                                                                              
3.49 µs ± 96.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: %%timeit 
   ...: U.narrow(-2, 0, 50) 
   ...:  
   ...:                                                                                                              
2.11 µs ± 88.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
Using narrow now.